File: ios/Classes/models/map/FCPMapTemplate.swift
//
//  FCPMapTemplate.swift
//  flutter_carplay
//
//  Created by Oğuzhan Atalay on 21.08.2021.
//

import CarPlay
import heresdk
import MapKit

/// A custom CarPlay map template with additional customization options.
@available(iOS 14.0, *)
class FCPMapTemplate: NSObject {
    // MARK: Properties

    /// The super template object representing the CarPlay map template.
    private(set) var _super: CPMapTemplate?

    /// The view controller associated with the map template.
    private(set) var viewController: UIViewController?

    /// The unique identifier for the map template.
    private(set) var elementId: String

    /// The title displayed on the map template.
    private var title: String?

    /// The map buttons to be displayed on the map template.
    private var mapButtons: [FCPMapButton]

    /// The leading navigation bar buttons for the map template.
    private var leadingNavigationBarButtons: [FCPBarButton]

    /// The trailing navigation bar buttons for the map template.
    private var trailingNavigationBarButtons: [FCPBarButton]

    /// A boolean value indicating whether the navigation bar is automatically hidden.
    private var automaticallyHidesNavigationBar: Bool = false

    /// A boolean value indicating whether buttons are hidden with the navigation bar.
    private var hidesButtonsWithNavigationBar: Bool = false

    /// Navigation session used to manage the upcomingManeuvers and  arrival estimation details
    var navigationSession: CPNavigationSession?

    /// Get the `FCPMapViewController` associated with the map template.
    var fcpMapViewController: FCPMapViewController? {
        return viewController as? FCPMapViewController
    }

    // MARK: Initializer

    /// Initializes a new instance of `FCPMapTemplate` with the specified configuration.
    ///
    /// - Parameter obj: A dictionary containing the configuration parameters for the map template.
    init(obj: [String: Any]) {
        guard let elementId = obj["_elementId"] as? String else {
            fatalError("[FCPMapTemplate] Missing required property: _elementId.")
        }

        self.elementId = elementId
        title = obj["title"] as? String
        automaticallyHidesNavigationBar = obj["automaticallyHidesNavigationBar"] as? Bool ?? false
        hidesButtonsWithNavigationBar = obj["hidesButtonsWithNavigationBar"] as? Bool ?? false

        mapButtons = (obj["mapButtons"] as? [[String: Any]] ?? []).map {
            FCPMapButton(obj: $0)
        }

        leadingNavigationBarButtons = (obj["leadingNavigationBarButtons"] as? [[String: Any]] ?? []).map {
            FCPBarButton(obj: $0)
        }

        trailingNavigationBarButtons = (obj["trailingNavigationBarButtons"] as? [[String: Any]] ?? []).map {
            FCPBarButton(obj: $0)
        }

        /// Initialize the view controller.
        viewController = FCPMapViewController(nibName: "FCPMapViewController", bundle: Bundle(for: FCPMapViewController.self))
    }

    // MARK: Getter

    /// Gets the CarPlay map template object based on the configured parameters.
    var get: CPMapTemplate {
        let mapTemplate = CPMapTemplate()
        mapTemplate.setFCPObject(self)

        var mButtons: [CPMapButton] = []
        for button in mapButtons {
            mButtons.append(button.get)
        }

        var lBButtons: [CPBarButton] = []
        for button in leadingNavigationBarButtons {
            lBButtons.append(button.get)
        }

        var tBButtons: [CPBarButton] = []
        for button in trailingNavigationBarButtons {
            tBButtons.append(button.get)
        }

        mapTemplate.mapButtons = mButtons
        mapTemplate.leadingNavigationBarButtons = lBButtons
        mapTemplate.trailingNavigationBarButtons = tBButtons
        mapTemplate.automaticallyHidesNavigationBar = automaticallyHidesNavigationBar
        mapTemplate.hidesButtonsWithNavigationBar = hidesButtonsWithNavigationBar
        mapTemplate.mapDelegate = self

        _super = mapTemplate
        return mapTemplate
    }

    /// Updates the properties of the map template.
    ///
    /// - Parameters:
    ///   - title: The new title text.
    ///   - automaticallyHidesNavigationBar: A boolean value indicating whether the navigation bar is automatically hidden.
    ///   - hidesButtonsWithNavigationBar: A boolean value indicating whether buttons are hidden with the navigation bar.
    ///   - mapButtons: The new array of map buttons.
    ///   - leadingNavigationBarButtons: The new array of leading navigation bar buttons.
    ///   - trailingNavigationBarButtons: The new array of trailing navigation bar buttons.
    public func update(title: String?, automaticallyHidesNavigationBar: Bool?, hidesButtonsWithNavigationBar: Bool?, mapButtons: [FCPMapButton]?, leadingNavigationBarButtons: [FCPBarButton]?, trailingNavigationBarButtons: [FCPBarButton]?) {
        if let _title = title {
            self.title = _title
        }

        if let _hidesButtonsWithNavigationBar = hidesButtonsWithNavigationBar {
            self.hidesButtonsWithNavigationBar = _hidesButtonsWithNavigationBar
            _super?.hidesButtonsWithNavigationBar = _hidesButtonsWithNavigationBar
        }

        if let _automaticallyHidesNavigationBar = automaticallyHidesNavigationBar {
            self.automaticallyHidesNavigationBar = _automaticallyHidesNavigationBar
            _super?.automaticallyHidesNavigationBar = _automaticallyHidesNavigationBar
        }

        if let _mapButtons = mapButtons {
            self.mapButtons = _mapButtons
            _super?.mapButtons = _mapButtons.map { $0.get }
        }

        if let _leadingNavigationBarButtons = leadingNavigationBarButtons {
            self.leadingNavigationBarButtons = _leadingNavigationBarButtons
            _super?.leadingNavigationBarButtons = _leadingNavigationBarButtons.map {
                $0.get
            }
        }

        if let _trailingNavigationBarButtons = trailingNavigationBarButtons {
            self.trailingNavigationBarButtons = _trailingNavigationBarButtons
            _super?.trailingNavigationBarButtons = _trailingNavigationBarButtons.map {
                $0.get
            }
        }
        _super?.setFCPObject(self)
    }
}

// MARK: - Trip

extension FCPMapTemplate {
    /// Show trip previews
    func showTripPreviews(trips: [FCPTrip], selectedTrip: FCPTrip?, textConfiguration: FCPTripPreviewTextConfiguration?) {
        let cpTrips = trips.map { $0.get }
        _super?.showTripPreviews(cpTrips, selectedTrip: selectedTrip?.get,
                                 textConfiguration: textConfiguration?.get)
    }

    /// Hide trip previews.
    func hideTripPreviews() {
        _super?.hideTripPreviews()
    }

    /// Starts the navigation
    /// - Parameter trip: The trip to start navigation
    func startNavigation(trip: CPTrip) {
        if navigationSession != nil {
            navigationSession?.finishTrip()
            navigationSession = nil
        }

        hideTripPreviews()
        navigationSession = _super?.startNavigationSession(for: trip)
        fcpMapViewController?.startNavigation(trip: trip)
    }

    /// Stops the navigation
    func stopNavigation() {
        navigationSession?.finishTrip()
        navigationSession = nil

        fcpMapViewController?.stopNavigation()
    }

    func showPanningInterface(animated: Bool) {
        _super?.showPanningInterface(animated: animated)
    }

    func dismissPanningInterface(animated: Bool) {
        _super?.dismissPanningInterface(animated: animated)
    }
}

// MARK: - CPMapTemplateDelegate

extension FCPMapTemplate: CPMapTemplateDelegate {
    func mapTemplate(_: CPMapTemplate, startedTrip trip: CPTrip, using _: CPRouteChoice) {
        let originCoordinate = trip.origin.placemark.coordinate
        let destinationCoordinate = trip.destination.placemark.coordinate

        DispatchQueue.main.async {
            FCPStreamHandlerPlugin.sendEvent(type: FCPChannelTypes.onNavigationStartedFromCarPlay, data: ["sourceLatitude": originCoordinate.latitude, "sourceLongitude": originCoordinate.longitude, "destinationLatitude": destinationCoordinate.latitude, "destinationLongitude": destinationCoordinate.longitude])
        }
    }

    /// Called when the panning interface is shown
    /// - Parameter _:  The map template
    func mapTemplateDidShowPanningInterface(_: CPMapTemplate) {}

    /// Called when the panning interface is dismissed
    /// - Parameter _: The map template
    func mapTemplateDidDismissPanningInterface(_: CPMapTemplate) {}

    func mapTemplate(_: CPMapTemplate, panWith direction: CPMapTemplate.PanDirection) {
        fcpMapViewController?.panInDirection(direction)
    }
}

@available(iOS 14.0, *)
extension FCPMapTemplate: FCPRootTemplate {}

@available(iOS 14.0, *)
extension FCPMapTemplate: FCPTemplate {}

File: android/src/main/kotlin/com/oguzhnatly/flutter_carplay/models/map/FCPMapTemplate.kt
package com.oguzhnatly.flutter_carplay.models.map

import androidx.car.app.AppManager
import androidx.car.app.OnDoneCallback
import androidx.car.app.SurfaceCallback
import androidx.car.app.model.Action
import androidx.car.app.model.ActionStrip
import androidx.car.app.navigation.NavigationManager
import androidx.car.app.navigation.NavigationManagerCallback
import androidx.car.app.navigation.model.NavigationTemplate
import androidx.car.app.serialization.Bundleable
import com.oguzhnatly.flutter_carplay.AndroidAutoService
import com.oguzhnatly.flutter_carplay.Bool
import com.oguzhnatly.flutter_carplay.CPMapTemplate
import com.oguzhnatly.flutter_carplay.CPTrip
import com.oguzhnatly.flutter_carplay.FCPRootTemplate
import com.oguzhnatly.flutter_carplay.models.button.FCPBarButton


/**
 * A custom Android Auto map template with additional customization options.
 *
 * @param obj A dictionary containing the configuration parameters for the map template.
 */
class FCPMapTemplate(obj: Map<String, Any>) : FCPRootTemplate() {

    /// The super template object representing the CarPlay map template.
    lateinit var _super: CPMapTemplate

    /// The view controller associated with the map template.
    var viewController: SurfaceCallback
        private set

    /// The title displayed on the map template.
    private var title: String?

    /// The map buttons to be displayed on the map template.
    private var mapButtons: List<FCPMapButton>

    /// The leading navigation bar buttons for the map template.
    private var leadingNavigationBarButtons: List<FCPBarButton>

    /// The trailing navigation bar buttons for the map template.
    private var trailingNavigationBarButtons: List<FCPBarButton>

//    /// The dashboard buttons to be displayed on the CarPlay dashboard.
//    private var dashboardButtons: List<FCPDashboardButton>

    /// A boolean value indicating whether the navigation bar is automatically hidden.
    private var automaticallyHidesNavigationBar: Bool

    /// A boolean value indicating whether buttons are hidden with the navigation bar.
    private var hidesButtonsWithNavigationBar: Bool

    /// A boolean value indicating whether the map is in panning mode.
    var isPanningInterfaceVisible: Bool = false

    /// Navigation session used to manage the upcomingManeuvers and  arrival estimation details
    var navigationSession: NavigationManager? = null

    /// Get the `FCPMapViewController` associated with the map template.
    val fcpMapViewController: FCPMapViewController?
        get() = viewController as? FCPMapViewController

    init {
        val elementIdValue = obj["_elementId"] as? String
        assert(elementIdValue != null) {
            "Missing required keys in dictionary for FCPMapTemplate initialization."
        }
        elementId = elementIdValue!!
        title = obj["title"] as? String
        automaticallyHidesNavigationBar = obj["automaticallyHidesNavigationBar"] as? Bool ?: false
        hidesButtonsWithNavigationBar = obj["hidesButtonsWithNavigationBar"] as? Bool ?: false
        isPanningInterfaceVisible = obj["isPanningInterfaceVisible"] as? Bool ?: false

        mapButtons = (obj["mapButtons"] as? List<Map<String, Any>>)?.map {
            FCPMapButton(it)
        } ?: emptyList()

//        dashboardButtons = (obj["dashboardButtons"] as? List<Map<String, Any>>)?.map {
//            FCPDashboardButton(it)
//        } ?: emptyList()

        leadingNavigationBarButtons =
            (obj["leadingNavigationBarButtons"] as? List<Map<String, Any>>)?.map {
                FCPBarButton(it)
            }
                ?: emptyList()
        trailingNavigationBarButtons =
            (obj["trailingNavigationBarButtons"] as? List<Map<String, Any>>)?.map {
                FCPBarButton(it)
            }
                ?: emptyList()

        // Initialize the view controller.
        viewController = FCPMapViewController()

        AndroidAutoService.session?.carContext?.getCarService(AppManager::class.java)
            ?.setSurfaceCallback(viewController)

        navigationSession =
            AndroidAutoService.session?.carContext?.getCarService(NavigationManager::class.java)

        navigationSession?.setNavigationManagerCallback(object : NavigationManagerCallback {
            /**
             * Overrides the `onStopNavigation` function and calls the `super.onStopNavigation()`
             * method to perform any necessary cleanup. Then, calls the `stopNavigation()`
             * method to stop the navigation.
             *
             * @return void
             */
            override fun onStopNavigation() {
                stopNavigation()
            }
        })
    }

    /** Gets the Android Auto map template object based on the configured parameters. */
    override fun getTemplate(): CPMapTemplate {
        val mapTemplate = NavigationTemplate.Builder()
        if (mapButtons.isNotEmpty()) {
            val actionStrip = ActionStrip.Builder()
            for (button in mapButtons) {
                actionStrip.addAction(button.getTemplate())
            }
            actionStrip.addAction(Action.PAN)
            mapTemplate.setMapActionStrip(actionStrip.build())
        }

        if (leadingNavigationBarButtons.isNotEmpty() || trailingNavigationBarButtons.isNotEmpty()) {
            val actionStrip = ActionStrip.Builder()
            for (button in leadingNavigationBarButtons) {
                actionStrip.addAction(button.getTemplate())
            }
            for (button in trailingNavigationBarButtons) {
                actionStrip.addAction(button.getTemplate())
            }
            mapTemplate.setActionStrip(actionStrip.build())
        }

        mapTemplate.setPanModeListener { isPanningInterfaceVisible = it }

//        mapTemplate.automaticallyHidesNavigationBar = automaticallyHidesNavigationBar
//        mapTemplate.hidesButtonsWithNavigationBar = hidesButtonsWithNavigationBar
//        mapTemplate.mapDelegate = self

        _super = mapTemplate.build()
        return _super
    }

    /**
     * Updates the properties of the map template.
     *
     * @param title The new title text.
     * @param automaticallyHidesNavigationBar A boolean value indicating whether the navigation bar is automatically hidden.
     * @param hidesButtonsWithNavigationBar A boolean value indicating whether buttons are hidden with the navigation bar.
     * @param mapButtons The new array of map buttons.
     * @param leadingNavigationBarButtons The new array of leading navigation bar buttons.
     * @param trailingNavigationBarButtons The new array of trailing navigation bar buttons.
     */
    fun update(
        title: String?,
        automaticallyHidesNavigationBar: Bool?,
        hidesButtonsWithNavigationBar: Bool?,
        isPanningInterfaceVisible: Bool?,
        mapButtons: List<FCPMapButton>?,
        leadingNavigationBarButtons: List<FCPBarButton>?,
        trailingNavigationBarButtons: List<FCPBarButton>?,
    ) {
        title?.let { this.title = it }
        automaticallyHidesNavigationBar?.let { this.automaticallyHidesNavigationBar = it }
        hidesButtonsWithNavigationBar?.let { this.hidesButtonsWithNavigationBar = it }
        isPanningInterfaceVisible?.let { this.isPanningInterfaceVisible = it }
        mapButtons?.let { this.mapButtons = it }
        leadingNavigationBarButtons?.let { this.leadingNavigationBarButtons = it }
        trailingNavigationBarButtons?.let { this.trailingNavigationBarButtons = it }

        onInvalidate()
    }
}

/**
 * Show trip previews
 *
 * @param trips The array of trips to show
 * @param selectedTrip The selected trip
 * @param textConfiguration The text configuration
 */
fun FCPMapTemplate.showTripPreviews(
    trips: List<FCPTrip>,
    selectedTrip: FCPTrip?,
//    textConfiguration: FCPTripPreviewTextConfiguration?,
) {
    val cpTrips = trips.map { it.getTemplate() }
//    _super?.showTripPreviews(cpTrips, selectedTrip: selectedTrip?. get,
//    textConfiguration: textConfiguration?.get)
}

/** Hide trip previews. */
fun FCPMapTemplate.hideTripPreviews() {
//    _super?.hideTripPreviews()
}

/**
 * Starts the navigation
 *
 * @param trip The trip to start navigation
 */
fun FCPMapTemplate.startNavigation(trip: CPTrip) {
//    if (navigationSession != null) {
//        navigationSession?.navigationEnded()
//        navigationSession = null
//    }

    hideTripPreviews()
//    navigationSession = _super?.startNavigationSession(for: trip)

//    if # available(iOS 15.4, *) {
//        navigationSession?.pauseTrip(for:.loading, description: "", turnCardColor: .systemGreen)
//    } else {
//        navigationSession?.pauseTrip(for:.loading, description: "")
//    }

    navigationSession?.updateTrip(trip)
    fcpMapViewController?.startNavigation(trip)
    navigationSession?.navigationStarted()
}

/** Stops the navigation. */
fun FCPMapTemplate.stopNavigation() {
    navigationSession?.navigationEnded()
//    navigationSession = null

    fcpMapViewController?.stopNavigation()
}

/**
 * Pans the camera in the specified direction
 *
 * @param animated A boolean value indicating whether the transition should be animated
 */
fun FCPMapTemplate.showPanningInterface() {
    _super.panModeDelegate?.sendPanModeChanged(true, object : OnDoneCallback {
        override fun onSuccess(response: Bundleable?) {
            print("pan mode enable success: $response")
        }

        override fun onFailure(response: Bundleable) {
            print("pan mode enable failed: $response")

        }
    })
}

/**
 * Dismisses the panning interface
 *
 * @param animated A boolean value indicating whether the transition should be animated
 */
fun FCPMapTemplate.dismissPanningInterface() {
    _super.panModeDelegate?.sendPanModeChanged(false, object : OnDoneCallback {
        override fun onSuccess(response: Bundleable?) {
            print("pan mode enable success: $response")
        }

        override fun onFailure(response: Bundleable) {
            print("pan mode enable failed: $response")

        }
    })
}


//extension FCPMapTemplate: CPMapTemplateDelegate {
//    /// Called when the map template has started a trip
//    /// - Parameter
//    ///   - mapTemplate: The map template
//    ///   - trip: The trip that was started
//    ///   - routeChoice: The route choice
//    func mapTemplate (_: CPMapTemplate, startedTrip trip: CPTrip, using _: CPRouteChoice) {
//        let originCoordinate = trip . origin . placemark . coordinate
//                let destinationCoordinate = trip . destination . placemark . coordinate
//
//                DispatchQueue.main.async {
//                    FCPStreamHandlerPlugin.sendEvent(type: FCPChannelTypes. onNavigationStartedFromCarplay, data: ["sourceLatitude": originCoordinate.latitude, "sourceLongitude": originCoordinate.longitude, "destinationLatitude": destinationCoordinate.latitude, "destinationLongitude": destinationCoordinate.longitude])
//                }
//    }
//
//    /// Called when the panning interface is shown
//    /// - Parameter mapTemplate: The map template
//    func mapTemplateDidShowPanningInterface (_: CPMapTemplate) {
//        fcpMapViewController?.hideSubviews()
//        fcpMapViewController?.mapController?.navigationHelper.stopCameraTracking()
//    }
//
//    /// Called when the panning interface is dismissed
//    /// - Parameter mapTemplate: The map template
//    func mapTemplateDidDismissPanningInterface (_: CPMapTemplate) {
//        fcpMapViewController?.showSubviews()
//        fcpMapViewController?.mapController?.navigationHelper.startCameraTracking()
//    }
//
//    /// Called when the map template is panning
//    /// - Parameters:
//    ///   - maptemplate: The map template
//    ///   - direction: The direction of the panning
//    func mapTemplate (_: CPMapTemplate, panWith direction: CPMapTemplate.PanDirection) {
//        fcpMapViewController?.panInDirection(direction)
//    }
//}
